src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1288,113 +1300,87 @@ object Desugar {
             * This case handles for loops with a range clause of the form:
             *
             * <invariants>
-            * for i, j := range x {
+            * // i0 here is an int variable starting from 0 and used at the
+            * // encoded loop condition
+            * for i, j := range x with i0 {
             *     body
             * }
             *
             * In the case where x is a slice or array the following code is produced. Note
             * that everything is in a new block so it can shadow variables with the same
             * name declared outside. This is also the go behaviour.
             *
-            * var i int = 0
-            * var i0 int = 0 // since 'i' can change in the iteration we store the true index in i0
-            * var j elem(x) // [v] the type of the elements of x 
-            *
-            * c := x // save the value of the slice/array since changing it doesn't change the iteration
-            *
-            * if (0 <= i && i < len(c)) { // [v]
-            *     j = c[i]
-            * }
-            *
-            * invariant len(c) > 0 ==> 0 <= i0 && i0 <= len(c) // these invariants do not
-            * invariant len(c) > 0 ==> i == i0                 // require access to anything so they
-            * invariant len(c) > 0 ==> 0 <= i && i <= len(c)   // are added before the user invariants
-            * <invariants were i is replaced with f(i, len(c) > 0) and j with f(j, len(c) > 0)>
-            * invariant len(c) > 0 ==> (0 <= i && i < len(c) ==> j == c[i]) // [v]
-            * for i < len(c) {
+            * if (len(x) == 0) {
```

Comment:
How do you guarantee here that `x` is evaluated, at most, once?

Reply:
I have updated the comment to show that only c is used. Also, the length of the slice/array is also saved in the beginning because it might change but Go retains the original length